src/Microsoft.Azure.SignalR.AspNet/OwinExtensions.cs

```diff
@@ -30,7 +39,62 @@ public static void RunAzureSignalR(this IAppBuilder builder)
 
         public static void RunAzureSignalR(this IAppBuilder builder, HubConfiguration configuration)
         {
-            throw new NotImplementedException();
+            RunAzureSignalRCore(builder, configuration);
+        }
+
+        private static void RunAzureSignalRCore(IAppBuilder builder, HubConfiguration configuration)
+        {
+            // Replace default HubDispatcher with a custom one, which has its own negotiation logic
+            // https://github.com/SignalR/SignalR/blob/dev/src/Microsoft.AspNet.SignalR.Core/Hosting/PersistentConnectionFactory.cs#L42
+            var hubDispatcher = new ServiceHubDispatcher(configuration);
+            configuration.Resolver.Register(typeof(PersistentConnection), () => hubDispatcher);
+            builder.RunSignalR(typeof(PersistentConnection), configuration);
+
+            RegisterServiceObjects(configuration);
+
+            var hubs = GetAvailableHubNames(configuration);
+            if (hubs?.Count > 0)
+            {
+                // Start the server->service connection asynchronously 
+                _ = new ConnectionFactory(hubs, configuration).StartAsync();
+            }
+            else
+            {
+                // TODO: log something
+            }
+        }
+
+        private static void RegisterServiceObjects(HubConfiguration configuration)
+        {
+            // share the same object all through
+            var serviceOptions = new OptionsWrapper<ServiceOptions>(new ServiceOptions
```

Comment:
Options.Create(new ServiceOptions(...))

Reply:
Wait, do you even need the Options at all?

Reply:
I need an interface wrapper for `ServiceOptions` to DI, as I don't want to create an interface for `ServiceOptions`, (because Ninject expects interfaces...) so I reused the `IOptions`.